### PR TITLE
feat: support custom storage for every table(initial)

### LINKED ISSUE #919: Custom storage type for every table.
### What problem does the new feature solve?

The GreptimeDB already supports storing data in the local file system, AWS S3 and Aliyun Oss, thanks to OpenDAL.

But this storage configuration is global. You can't have a table in the local file system and the other on S3.

I think we can make this configuration also present in table engine options. For example, you can create a table and specify its storage type:

```sql
create table monitor(CPU double, ts timestamp time index) with(regions=1, storage=s3);
```

If the SQL doesn't set the `storage` option, then using the global configuration.

### What does the feature do?

Describe as above.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by NiwakaDev**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR addresses a part of #919. 

Here's an example of using this feature.

```sql
create table foo(a string, ts TIMESTAMP TIME INDEX, ~)with(storage='Gcs');
```

Note that this PR doesn't implement:
* custom storage for region failover.
* custom storage for file engine.
## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

Not close #919 yet.

### CODE REVIEW

**Comment by NiwakaDev** [comment]
I rebased the `develop` branch, and then `rustc` says:

```bash
error: use of deprecated method `chrono::DateTime::<Tz>::timestamp_nanos`: use `timestamp_nanos_opt()` instead
  --> src/common/time/src/timestamp.rs:66:41
   |
66 |             TimeUnit::Nanosecond => now.timestamp_nanos(),
   |                                         ^^^^^^^^^^^^^^^
   |
   = note: `-D deprecated` implied by `-D warnings`

error: use of deprecated method `chrono::TimeZone::datetime_from_str`: use `DateTime::parse_from_str` instead
   --> src/common/time/src/timestamp.rs:301:29
    |
301 |         if let Ok(ts) = Utc.datetime_from_str(s, "%Y-%m-%d %H:%M:%S%.fZ") {
    |                             ^^^^^^^^^^^^^^^^^
```

**Comment by evenyag** [comment]
@NiwakaDev Maybe caused by a newer opendal version. Our `Cargo.lock` file uses opendal `0.40.0`.

File: src/common/datasource/src/object_store.rs [hunk 1]
```diff
@@ -84,6 +85,86 @@ pub fn handle_windows_path(url: &str) -> Option<String> {
...
+        }
+    }
+}
+
+impl StorageType {
+    pub fn build_storage_types_based_on_env() -> Vec<StorageType> {
+        let mut storage_types = Vec::with_capacity(4);
+        storage_types.push(StorageType::File);
+        if let Ok(bucket) = env::var("GT_S3_BUCKET") {
+            if !bucket.is_empty() {
```
Comment (NiwakaDev) [inline src/common/datasource/src/object_store.rs:116]: https://github.com/GreptimeTeam/greptimedb/blob/3217b56cc10eeb7f334ee08ecb72b0a0557388f5/.github/workflows/develop.yml#L178-L184

Settings for S3 reside, but it seems like bucket is empty.

Comment (evenyag) [reply src/common/datasource/src/object_store.rs:116]: The `S3_BUCKET` is only available in this repo's secrets. A PR from a forked repo can't get this secret. So we run the test in the CI after the PR is merged.

**Comment by evenyag** [comment]
This feature impacts many modules and contains breaking changes. I'd like to merge it gradually (in several PRs) so we can review and test it carefully. Could you please separate it into smaller PRs? e.g.
- A PR to add the `ObjectStoreManager`
- A PR to support the `ObjectStoreManager` in mito2 (Now the manager only contains one object store)
- Support specifying storage in SQL and some tests
- More tests

File: config/datanode.example.toml [hunk 2]
```diff
@@ -43,9 +43,10 @@ sync_write = false
 [storage]
 # The working home directory.
 data_home = "/tmp/greptimedb/"
+# global_store = "File"
+# Storage type.
+[[storage.store]]
```
Comment (evenyag) [inline config/datanode.example.toml:48]: This is a breaking change. `storage.store` is a little odd. We'd better find out a new name or revisit the storage config later.

**Comment by NiwakaDev** [comment]
> This feature impacts many modules and contains breaking changes. I'd like to merge it gradually (in several PRs) so we can review and test it carefully. Could you please separate it into smaller PRs?

Sorry for submitting the learge PR. Your point is correct😅. I opened a new PR(#2621), and please review it.

**Pull request closed without merging**